gpii/node_modules/userListeners/src/listeners.js

```diff
@@ -0,0 +1,163 @@
+        "onStopListeners": null
+    },
+    listeners: {
+        "onCreate.startListeners": "{that}.events.onStartListeners"
+    }
+});
+
+// A user listener.
+fluid.defaults("gpii.userListener", {
+    gradeNames: ["fluid.component"],
+    events: {
+        // A token has arrived.
+        "onTokenArrive": null,
+        // A token has removed.
+        "onTokenRemove": null
+    },
+    invokers: {
+        // Start the listener, returning a Promise
+        startListener: "fluid.notImplemented",
+        // Stop the listener, returning a Promise
+        stopListener: "fluid.notImplemented",
+        // Parse a Buffer containing an NDEF message, returning a string.
+        decodeNDEF: {
+            funcName: "gpii.userListeners.decodeNDEF",
+            args: [
+                "{arguments}.0" // NDEF buffer
+            ]
+        }
+    },
+    members: {
+        // Set to true to call proximityTriggered, otherwise login/logout.
+        proximity: false,
+        listenerName: "@expand:fluid.identity({that}.typeName)"
+    },
+    listeners: {
+        // Start/stop this listener
+        "{userListeners}.events.onStartListeners":  "{that}.startListener",
+        "{userListeners}.events.onStopListeners": "{that}.stopListener",
+
+        "onTokenArrive.callFlowManager": {
+            funcName: "gpii.userListeners.tokenArrived",
+            args: [
+                "{that}",
+                "{arguments}.1" // token
+            ]
+        },
+        "onTokenRemove.callFlowManager": {
+            funcName: "gpii.userListeners.tokenRemoved",
+            args: [
+                "{that}",
+                "{arguments}.1" // token
+            ]
+        }
+    }
+});
+
+/**
+ * Handles the onTokenArrive event.
+ *
+ * It calls the "login" action for non-proximity devices, otherwise "proximityTriggered".
+ *
+ * @param that {Component} An instance of gpii.userListener.
+ * @param token {string} The token from the user listener.
+ */
+gpii.userListeners.tokenArrived = function (that, token) {
+    fluid.log(that.listenerName + " token arrived: " + token);
+
+    var action = that.proximity ? "proximityTriggered" : "login";
+    request("http://localhost:8081/user/" + token + "/" + action);
+};
+
+/**
+ * Handles the onTokenRemove event.
+ *
+ * It calls the "logout" action for non-proximity devices, otherwise it does nothing.
+ *
+ * @param that {Component} An instance of gpii.userListener.
+ * @param token {string} The token from the user listener.
+ */
+gpii.userListeners.tokenRemoved = function (that, token) {
+    fluid.log(that.listenerName + " token removed: " + token);
+
+    if (!that.proximity) {
+        request("http://localhost:8081/user/" + token + "/logout");
+    }
+};
+
+/**
+ * Decodes an NDEF message (the tag from a card), returning the value of the first text record.
+ *
+ * @param tagBuffer
+ * @return {string} The text value, or null if none.
+ */
+gpii.userListeners.decodeNDEF = function (tagBuffer) {
```

Comment:
Wouldn't this make more sense to have in pcsc.js? Since it's (from my understanding) only relevant to that listener?

Reply:
Yes, not sure why I put it in there.